File: src/services/sftp.ts
import { invoke } from '@tauri-apps/api/core'

/**
 * 统一包装 invoke 错误,生成面向用户可读的错误信息。
 * 返回类型为 never,调用方在 catch 块中调用后 TypeScript 会自动收窄类型。
 */
function wrapInvokeError(operation: string, error: unknown): never {
  const message = error instanceof Error ? error.message : String(error)
  throw new Error(`[SFTP] ${operation} 失败: ${message}`)
}

export interface SftpEntry {
  /** 文件/目录名(不含父路径) */
  name: string
  /** 完整绝对路径 */
  path: string
  /** 是否是目录 */
  isDir: boolean
  /** 文件大小(字节),目录为 0 */
  size: number
  /** 修改时间(Unix timestamp 毫秒) */
  modified?: number
  /** POSIX 权限位(0o755 = 493) */
  permissions: number
  /** Unix owner uid(可选) */
  uid?: number
  /** Unix group gid(可选) */
  gid?: number
}

export async function sftpList(id: string, path: string): Promise<SftpEntry[]> {
  try {
    return await invoke('sftp_list', { id, path })
  } catch (error) {
    wrapInvokeError('sftp_list', error)
  }
}

export async function sftpRead(id: string, path: string): Promise<number[]> {
  try {
    return await invoke('sftp_read', { id, path })
  } catch (error) {
    wrapInvokeError('sftp_read', error)
  }
}

export async function sftpWrite(id: string, path: string, data: number[]): Promise<void> {
  try {
    return await invoke('sftp_write', { id, path, data })
  } catch (error) {
    wrapInvokeError('sftp_write', error)
  }
}

export async function sftpStat(id: string, path: string): Promise<SftpEntry> {
  try {
    return await invoke('sftp_stat', { id, path })
  } catch (error) {
    wrapInvokeError('sftp_stat', error)
  }
}

export async function sftpRemove(id: string, path: string): Promise<void> {
  try {
    return await invoke('sftp_remove', { id, path })
  } catch (error) {
    wrapInvokeError('sftp_remove', error)
  }
}

export async function sftpMkdir(id: string, path: string): Promise<void> {
  try {
    return await invoke('sftp_mkdir', { id, path })
  } catch (error) {
    wrapInvokeError('sftp_mkdir', error)
  }
}

export async function sftpRename(id: string, from: string, to: string): Promise<void> {
  try {
    return await invoke('sftp_rename', { id, from, to })
  } catch (error) {
    wrapInvokeError('sftp_rename', error)
  }
}

// ===== 流式传输(走 TransferManager,带 progress / status 事件) =====

export interface TransferFile {
  name: string
  size: number
  transferred: number
}

export type TransferDirection = 'upload' | 'download'
export type TransferStatus = 'queued' | 'running' | 'done' | 'failed' | 'cancelled'

export interface TransferTask {
  id: string
  sessionId: string
  direction: TransferDirection
  files: TransferFile[]
  status: TransferStatus
  totalBytes: number
  transferredBytes: number
  error?: string | null
}

export interface TransferProgress {
  transferId: string
  fileName: string
  transferred: number
  total: number
  direction: TransferDirection
}

export interface TransferStatusEvent {
  transferId: string
  sessionId: string
  direction: TransferDirection
  status: TransferStatus
  error?: string | null
}

export interface SftpLaunchInfo {
  mode: 'subsystem' | 'fallback_exec' | 'custom_exec'
  server_path?: string | null
  diagnostic?: string | null
}

/** 先确保 SFTP 通道已开启并注册到 TransferManager(后续传输复用同一通道) */
export async function sftpEnsureSession(id: string): Promise<SftpLaunchInfo | null> {
  try {
    return await invoke('sftp_ensure_session', { id })
  } catch (error) {
    wrapInvokeError('sftp_ensure_session', error)
  }
}

/** 启动流式上传,返回 transfer_id(用于 listen 进度/状态) */
export async function sftpStartUpload(
  id: string,
  localPaths: string[],
  remoteDir: string,
  speedLimit: number = 0
): Promise<string> {
  try {
    return await invoke('sftp_start_upload', { id, localPaths, remoteDir, speedLimit })
  } catch (error) {
    wrapInvokeError('sftp_start_upload', error)
  }
}

/** 启动流式下载,返回 transfer_id */
export async function sftpStartDownload(
  id: string,
  remotePaths: string[],
  localDir: string,
  speedLimit: number = 0
): Promise<string> {
  try {
    return await invoke('sftp_start_download', { id, remotePaths, localDir, speedLimit })
  } catch (error) {
    wrapInvokeError('sftp_start_download', error)
  }
}

/** 取消一个传输 */
export async function sftpCancelTransfer(id: string, transferId: string): Promise<void> {
  try {
    return await invoke('sftp_cancel_transfer', { id, transferId })
  } catch (error) {
    wrapInvokeError('sftp_cancel_transfer', error)
  }
}

/** 重试一个失败/已取消的传输 */
export async function sftpRetryTransfer(id: string, transferId: string): Promise<string> {
  try {
    return await invoke('sftp_retry_transfer', { id, transferId })
  } catch (error) {
    wrapInvokeError('sftp_retry_transfer', error)
  }
}

/** 动态设置传输任务的速度限制(bytes/s,0 表示不限) */
export async function sftpSetSpeedLimit(id: string, transferId: string, speedLimit: number): Promise<void> {
  try {
    return await invoke('sftp_set_speed_limit', { id, transferId, speedLimit })
  } catch (error) {
    wrapInvokeError('sftp_set_speed_limit', error)
  }
}

/** 列出某 session 的所有传输任务 */
export async function sftpListTransfers(id: string): Promise<TransferTask[]> {
  try {
    return await invoke('sftp_list_transfers', { id })
  } catch (error) {
    wrapInvokeError('sftp_list_transfers', error)
  }
}

/** 把 parent + name 拼成 child 路径(保证中间有 /) */
export function joinPath(parent: string, name: string): string {
  if (parent === '/' || parent === '') return `/${name}`
  if (parent.endsWith('/')) return `${parent}${name}`
  return `${parent}/${name}`
}

/** 返回 parent 路径(根目录时返回 '/') */
export function parentPath(path: string): string {
  if (!path || path === '/') return '/'
  const i = path.lastIndexOf('/')
  if (i <= 0) return '/'
  return path.slice(0, i)
}

/** 字节数格式化为可读字符串 */
export function formatSize(bytes: number): string {
  if (bytes === 0) return '0 B'
  const units = ['B', 'KB', 'MB', 'GB', 'TB']
  const k = 1024
  const i = Math.floor(Math.log(bytes) / Math.log(k))
  return `${(bytes / Math.pow(k, i)).toFixed(i === 0 ? 0 : 1)} ${units[i]}`
}

File: src-tauri/src/commands/sftp.rs
use crate::commands::ssh::SshManager;
use crate::sftp::transfer::TransferManager;
use crate::sftp::TransferTask;
use crate::ssh::session::SftpLaunchInfo;
use crate::ssh::sftp::SftpEntry;
use std::sync::Arc;
use tauri::State;
use tokio::sync::Mutex;

fn map_err<E: std::fmt::Display>(e: E) -> String {
    format!("{}", e)
}

/// sftp_read 单次读取上限(整文件进内存 + IPC 传输)。超出请走 TransferManager 分块下载。
const MAX_SFTP_READ_BYTES: u64 = 1024 * 1024;
/// sftp_write 单次写入上限。超出请走 TransferManager 分块上传。
const MAX_SFTP_WRITE_BYTES: usize = 2 * 1024 * 1024;

/// 从 sessions map 中取出 Arc,立即释放主锁。
/// 返回的 Arc 由调用方持有,保证 MutexGuard 有效。
macro_rules! get_session_arc {
    ($manager:expr, $id:expr) => {{
        let sessions = $manager.sessions.lock().await;
        let arc = sessions
            .get(&$id)
            .cloned()
            .ok_or_else(|| "Session not found".to_string())?;
        drop(sessions);
        arc
    }};
}

/// 读取远程目录
#[tauri::command]
pub async fn sftp_list(
    manager: State<'_, SshManager>,
    id: String,
    path: String,
) -> Result<Vec<SftpEntry>, String> {
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    let read_dir = session
        .with_browse_sftp(|sftp| {
            Box::pin(async move { sftp.read_dir(&path).await.map_err(map_err) })
        })
        .await?;

    let mut entries: Vec<SftpEntry> = Vec::new();
    for dir_entry in read_dir {
        let name = dir_entry.file_name();
        let metadata = dir_entry.metadata();
        let full_path = dir_entry.path();
        let is_dir = metadata.is_dir();
        let size = if is_dir {
            0
        } else {
            metadata.size.unwrap_or(0)
        };
        let modified = metadata.mtime.map(|t| (t as u64) * 1000);

        entries.push(SftpEntry {
            name,
            path: full_path,
            is_dir,
            size,
            modified,
            permissions: metadata.permissions.unwrap_or(0),
            uid: metadata.uid,
            gid: metadata.gid,
        });
    }

    // 目录在前,文件在后,字母序
    entries.sort_by(|a, b| match (a.is_dir, b.is_dir) {
        (true, false) => std::cmp::Ordering::Less,
        (false, true) => std::cmp::Ordering::Greater,
        _ => a.name.to_lowercase().cmp(&b.name.to_lowercase()),
    });

    Ok(entries)
}

/// 读文件内容(返回字节)。整文件进内存并经 IPC 传输,限制 1MB;
/// 更大的文件请走 TransferManager 分块下载(sftp_start_download)。
#[tauri::command]
pub async fn sftp_read(
    manager: State<'_, SshManager>,
    id: String,
    path: String,
) -> Result<Vec<u8>, String> {
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    session
        .with_browse_sftp(|sftp| {
            Box::pin(async move {
                let metadata = sftp.metadata(&path).await.map_err(map_err)?;
                let size = metadata.size.unwrap_or(0);
                if size > MAX_SFTP_READ_BYTES {
                    return Err(format!(
                        "[SFTP_FILE_TOO_LARGE] File is {} bytes, exceeding the {} bytes limit of sftp_read; use chunked download (sftp_start_download) instead",
                        size, MAX_SFTP_READ_BYTES
                    ));
                }
                sftp.read(&path).await.map_err(map_err)
            })
        })
        .await
}

/// 写文件(覆盖)。整文件进内存并经 IPC 传输,限制 2MB;
/// 更大的文件请走 TransferManager 分块上传(sftp_start_upload)。
#[tauri::command]
pub async fn sftp_write(
    manager: State<'_, SshManager>,
    id: String,
    path: String,
    data: Vec<u8>,
) -> Result<(), String> {
    if data.len() > MAX_SFTP_WRITE_BYTES {
        return Err(format!(
            "[SFTP_FILE_TOO_LARGE] Payload is {} bytes, exceeding the {} bytes limit of sftp_write; use chunked upload (sftp_start_upload) instead",
            data.len(),
            MAX_SFTP_WRITE_BYTES
        ));
    }
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    session
        .with_browse_sftp(|sftp| {
            Box::pin(async move { sftp.write(&path, &data).await.map_err(map_err) })
        })
        .await
}

/// 读文件元数据
#[tauri::command]
pub async fn sftp_stat(
    manager: State<'_, SshManager>,
    id: String,
    path: String,
) -> Result<SftpEntry, String> {
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    let meta_path = path.clone();
    let metadata = session
        .with_browse_sftp(|sftp| {
            Box::pin(async move { sftp.metadata(&meta_path).await.map_err(map_err) })
        })
        .await?;
    let name = std::path::Path::new(&path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or(&path)
        .to_string();
    Ok(SftpEntry {
        name,
        path: path.clone(),
        is_dir: metadata.is_dir(),
        size: metadata.size.unwrap_or(0),
        modified: metadata.mtime.map(|t| (t as u64) * 1000),
        permissions: metadata.permissions.unwrap_or(0),
        uid: metadata.uid,
        gid: metadata.gid,
    })
}

/// 删除文件(单文件)
#[tauri::command]
pub async fn sftp_remove_file(
    manager: State<'_, SshManager>,
    id: String,
    path: String,
) -> Result<(), String> {
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    session
        .with_browse_sftp(|sftp| {
            Box::pin(async move { sftp.remove_file(&path).await.map_err(map_err) })
        })
        .await
}

/// 删除目录(空目录)
#[tauri::command]
pub async fn sftp_remove_dir(
    manager: State<'_, SshManager>,
    id: String,
    path: String,
) -> Result<(), String> {
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    session
        .with_browse_sftp(|sftp| {
            Box::pin(async move { sftp.remove_dir(&path).await.map_err(map_err) })
        })
        .await
}

/// 删除文件或空目录(自动判断)
#[tauri::command]
pub async fn sftp_remove(
    manager: State<'_, SshManager>,
    id: String,
    path: String,
) -> Result<(), String> {
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    session
        .with_browse_sftp(|sftp| {
            Box::pin(async move {
                let metadata = sftp.metadata(&path).await.map_err(map_err)?;
                if metadata.is_dir() {
                    sftp.remove_dir(&path).await.map_err(map_err)?;
                } else {
                    sftp.remove_file(&path).await.map_err(map_err)?;
                }
                Ok(())
            })
        })
        .await
}

/// 创建目录
#[tauri::command]
pub async fn sftp_mkdir(
    manager: State<'_, SshManager>,
    id: String,
    path: String,
) -> Result<(), String> {
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    session
        .with_browse_sftp(|sftp| {
            Box::pin(async move { sftp.create_dir(&path).await.map_err(map_err) })
        })
        .await
}

/// 重命名
#[tauri::command]
pub async fn sftp_rename(
    manager: State<'_, SshManager>,
    id: String,
    from: String,
    to: String,
) -> Result<(), String> {
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    session
        .with_browse_sftp(|sftp| {
            Box::pin(async move { sftp.rename(&from, &to).await.map_err(map_err) })
        })
        .await
}

// ============== 流式传输(走 TransferManager,分块 + progress 事件) ==============

/// 打开一条 SFTP 通道并注册到 TransferManager(后续 sftp_start_*
/// 会复用它,避免每个文件都新开 SFTP channel)
#[tauri::command]
pub async fn sftp_ensure_session(
    manager: State<'_, SshManager>,
    transfer_manager: State<'_, TransferManager>,
    id: String,
) -> Result<Option<SftpLaunchInfo>, String> {
    // 已注册过就直接返回
    if transfer_manager.has_session(&id).await {
        tracing::info!("[sftp_ensure_session] session {} already registered", id);
        return Ok(None);
    }
    tracing::info!(
        "[sftp_ensure_session] opening SFTP channel for session {}",
        id
    );
    let session_arc = get_session_arc!(manager, id);
    let mut session = session_arc.lock().await;
    let (sftp, launch_info) = session.open_sftp_with_info().await.map_err(map_err)?;
    tracing::info!("[sftp_ensure_session] SFTP channel opened, registering to TransferManager");
    transfer_manager
        .register_sftp(id.clone(), Arc::new(Mutex::new(sftp)))
        .await;
    tracing::info!(
        "[sftp_ensure_session] session {} registered successfully",
        id
    );
    Ok(Some(launch_info))
}

/// 启动流式上传(分块 + progress 事件)
#[tauri::command]
pub async fn sftp_start_upload(
    transfer_manager: State<'_, TransferManager>,
    id: String,
    local_paths: Vec<String>,
    remote_dir: String,
    speed_limit: u64,
) -> Result<String, String> {
    transfer_manager
        .upload(&id, local_paths, remote_dir, speed_limit)
        .await
        .map_err(map_err)
}

/// 启动流式下载(分块 + progress 事件)
#[tauri::command]
pub async fn sftp_start_download(
    transfer_manager: State<'_, TransferManager>,
    id: String,
    remote_paths: Vec<String>,
    local_dir: String,
    speed_limit: u64,
) -> Result<String, String> {
    transfer_manager
        .download(&id, remote_paths, local_dir, speed_limit)
        .await
        .map_err(map_err)
}

/// 取消一个传输(设置 cancellation token,正在跑的循环检测到后退出;
/// 已暂停的任务直接落终态)
#[tauri::command]
pub async fn sftp_cancel_transfer(
    transfer_manager: State<'_, TransferManager>,
    _id: String,
    transfer_id: String,
) -> Result<(), String> {
    transfer_manager.cancel(&transfer_id).await;
    Ok(())
}

/// 暂停一个运行中的传输(保留断点偏移,可 resume 继续)
#[tauri::command]
pub async fn sftp_pause_transfer(
    transfer_manager: State<'_, TransferManager>,
    _id: String,
    transfer_id: String,
) -> Result<(), String> {
    transfer_manager.pause(&transfer_id).await;
    Ok(())
}

/// 继续一个已暂停的传输(从断点偏移续传)
#[tauri::command]
pub async fn sftp_resume_transfer(
    transfer_manager: State<'_, TransferManager>,
    _id: String,
    transfer_id: String,
) -> Result<(), String> {
    transfer_manager.resume(&transfer_id).await.map_err(map_err)
}

/// 动态修改传输速度限制(bytes/sec, 0 = 不限)
#[tauri::command]
pub async fn sftp_set_speed_limit(
    transfer_manager: State<'_, TransferManager>,
    _id: String,
    transfer_id: String,
    speed_limit: u64,
) -> Result<(), String> {
    transfer_manager
        .set_speed_limit(&transfer_id, speed_limit)
        .await;
    Ok(())
}

/// 重试失败的传输(创建新传输,从失败文件处继续)
#[tauri::command]
pub async fn sftp_retry_transfer(
    transfer_manager: State<'_, TransferManager>,
    _id: String,
    transfer_id: String,
) -> Result<String, String> {
    transfer_manager.retry(&transfer_id).await.map_err(map_err)
}

/// 列出某 session 的所有传输任务(给前端刷新/重连用)
#[tauri::command]
pub async fn sftp_list_transfers(
    transfer_manager: State<'_, TransferManager>,
    id: String,
) -> Result<Vec<TransferTask>, String> {
    Ok(transfer_manager.list_tasks(&id).await)
}
